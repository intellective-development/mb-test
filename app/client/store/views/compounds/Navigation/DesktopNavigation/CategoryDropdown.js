// @flow

import * as React from 'react';
import i18n from 'store/localization';
import _ from 'lodash';
import bindClassNames from 'shared/utils/bind_classnames';
import type { ContentModule } from 'store/business/content_module';

import CategoryDropdownContentModule, { CategoryDropdownContentModuleDivider } from './CategoryDropdownContentModule';
import { MBLayout, MBText } from '../../../elements';
import styles from './CategoryDropdown.scss';

const cn = bindClassNames(styles);

type NavigationDropdownProps = {
  content_modules: ContentModule[],
  selected_category_name: ?string,
  is_visible: boolean,
  is_available: boolean
}

class NavigationDropdown extends React.Component<NavigationDropdownProps> {
  renderDropdownElement(){
    const { is_available, selected_category_name, content_modules } = this.props;

    if (selected_category_name && !is_available){
      return <NavigationDropdownCategoryUnavailable category_name={selected_category_name} />;
    } else if (selected_category_name && !_.isEmpty(content_modules)){
      return <NavigationDropdownContentLayout content_modules={content_modules} />;
    } else {
      return null;
    }
  }

  render(){
    const { is_visible } = this.props;

    return (
      <div className={cn('cmDCategoryDropdown_Wrapper', {cmDCategoryDropdown_Wrapper__Open: is_visible})}>
        <div className={cn('cmDCategoryDropdown_Dropdown', {cmDCategoryDropdown__Dropdown_Open: is_visible})}>
          {this.renderDropdownElement()}
        </div>
      </div>
    );
  }
}

const NavigationDropdownCategoryUnavailable = ({category_name}) => {
  const unavailable_message = i18n.t(`ui.body.navigation.category_dropdown.${category_name}`);

  return (
    <MBText.P className={styles.cmDCategoryDropdown_UnavailableMessage}>{unavailable_message}</MBText.P>
  );
};

const NavigationDropdownContentLayout = ({content_modules}) => {
  const content = _.flatMap(content_modules, (content_module) => ([
    <CategoryDropdownContentModuleDivider
      module_type={content_module.module_type}
      key={`divider_${content_module.internal_name}`} />,
    <CategoryDropdownContentModule
      content_module={content_module}
      key={content_module.internal_name} />
  ]));

  return <MBLayout.StandardGrid no_padding className={styles.cmDCategoryDropdown_AvailableContainer}>{content}</MBLayout.StandardGrid>;
};

export default NavigationDropdown;
