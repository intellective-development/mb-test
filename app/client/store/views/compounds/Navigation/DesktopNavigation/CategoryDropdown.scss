@import "~client/store/views/style/variables";
@import "./_variables";

$dropdown_height: 256px; // used to animate the dropdown
$dropdown_shadow_height: 4px;
$module-spacing: $standard-gutter * 2;

.cmDCategoryDropdown_Wrapper {
  position: absolute;
  left: 0;
  right: 0;
  margin: 0;
  top: 100%; // stick it below the bottom of the container
  z-index: 10;
  border-top: $default-border;

  // this padding allows the dropdown's box shadow to show, despite the above overflow: hidden
  padding-bottom: $dropdown_shadow_height;
  overflow: hidden;
  visibility: hidden;
}

.cmDCategoryDropdown_Wrapper__Open {
  visibility: visible;
}

.cmDCategoryDropdown_Dropdown {
  background: $brand-white;
  margin-bottom: 0;
  box-shadow: 0 $dropdown_shadow_height/2 $dropdown_shadow_height 0 rgba(0, 0, 0, 0.33);
  transition: all 250ms ease-in; // animate out
  transform: translateY($dropdown_height * -1.2); // slide out a bit taller than normal height, just in case
  opacity: 0.75;
}

.cmDCategoryDropdown__Dropdown_Open {
  transition-timing-function: ease-out; // animate in
  transform: translateY(0);
  opacity: 1;
}

$short-content-module-count: 2;
$suppress-nth-child-offset: $short-content-module-count * 2 + 1;
.cmDCategoryDropdown_AvailableContainer {
  display: flex;
  padding: $standard-gutter * 2;
  padding-left: $standard-gutter * 2 + $full-logo-size;
  min-height: $dropdown_height;

  @media only screen and (max-width: $nav-contract-max-width) {
    padding-right: $standard-gutter;
    padding-left: $standard-gutter + $full-logo-size;

    // we suppress everything past the first two content modules
    .cmDCategoryDropdown_Divider:nth-child(n + #{$suppress-nth-child-offset}),
    .cmDCategoryDropdown_ContentModule:nth-child(n + #{$suppress-nth-child-offset}) {
      display: none;
    }
  }
}

.cmDCategoryDropdown_UnavailableMessage {
  @include font-size(15);

  color: $brand-error-red;
  padding: $standard-gutter*1.5 $standard-gutter;
  text-align: center;
}

/* Content Modules */

/* link_list */
$link-list-element-side-padding: 8px;

.cmDCategoryDropdown_LinkList_Container {
  flex: none;
}

.cmDCategoryDropdown_LinkList_Title {
  @include font-size(13);

  color: $brand-medium-grey;
  font-weight: $heavy-font-weight;
  text-transform: uppercase;
  letter-spacing: 0.125rem;
  padding-bottom: 1em;
}

.cmDCategoryDropdown_LinkList_ColGroup {
  display: flex;
  flex: none;
  margin-left: $link-list-element-side-padding * -1; // keep the links left aligned with the title
}

.cmDCategoryDropdown_LinkList_Col {
  display: flex;
  flex-direction: column;
  margin: 0;

  &:not(:last-child) { padding-right: $module-spacing; }
}

.cmDCategoryDropdown_LinkList_ElLink {
  @include font-size(15);

  display: block;
  width: 128px;
  line-height: 1.3em;
  padding: 0.25em $link-list-element-side-padding 0.2em;
  color: $brand-black;
  transition: background 150ms ease-in-out;

  &:hover,
  &:focus {
    color: $brand-black; // prevent color change
  }

  &:hover {
    background-color: $brand-pale-blue;
  }
}

.cmDCategoryDropdown_LinkList_ElLink__ShopAll {
  font-weight: $heavy-font-weight;
}

/* carousel */

.cmDCategoryDropdown_Carousel {
  flex: initial;
  align-self: flex-start;

  &:hover {
    opacity: 0.85;
  }
}

/* divider */

.cmDCategoryDropdown_Divider {
  padding: 0 $module-spacing;

  // we never want to show the first divider (it would be the leftmost element in the dropdown)
  &:first-of-type { display: none; }
}

.cmDCategoryDropdown_DividerLine {
  width: 1px;
  background-color: $brand-light-grey;
  height: 100%;
}
