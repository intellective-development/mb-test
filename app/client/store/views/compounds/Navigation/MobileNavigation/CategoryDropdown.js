// @flow
import * as React from 'react';
import _ from 'lodash';
import bindClassNames from 'shared/utils/bind_classnames';
import type { ContentLayout } from 'store/business/content_layout';
import type { ImageContent } from 'store/business/content_module';
import { NAVIGATION_CATEGORY_DEFAULTS } from '../constants';

import withContentLayout from '../../../compounds/ContentLayout';
import { MBLink, MBText, withUniqueId } from '../../../elements';
import styles from './CategoryDropdown.scss';

const cn = bindClassNames(styles);

type CategoryDropdownProps = {|
  is_showing: boolean,
  content_layout: ContentLayout
|};
export const CategoryDropdown = ({is_showing, content_layout}: CategoryDropdownProps) => {
  let category_links = NAVIGATION_CATEGORY_DEFAULTS;
  if (content_layout){
    // we assume that this content layout will only have one element, which contains the desired links
    category_links = _.get(content_layout, 'content[0].config.links');
  }

  return (
    <div className={cn('cmMCategoryDropdown_Wrapper', {cmMCategoryDropdown_Wrapper__DropdownVisible: is_showing})}>
      <ul className={cn('cmMCategoryDropdown', {cmMCategoryDropdown__Visible: is_showing})}>
        {category_links.map(el => (
          <CategoryElement
            key={el.name}
            name={el.name}
            icon_banner={el.icon_banner}
            action_url={el.action_url}
            disabled={!el.action_url} />
        ))}
      </ul>
    </div>
  );
};

type CategoryElementProps = {|
  action_url: string,
  name: 'wine' | 'beer' | 'liquor' | 'mixers' | 'gifts',
  icon_banner?: ImageContent,
  disabled?: boolean
|};
const CategoryElement = ({ action_url, name, icon_banner, disabled }: CategoryElementProps) => {
  return (
    <li>
      <MBLink.View
        disabled={disabled}
        className={cn('cmMCategoryDropdown_Element', {cmMCategoryDropdown_Element__Disabled: disabled})}
        href={action_url}>
        <CategoryElementIcon category_name={name} icon_banner={icon_banner} />
        <MBText.Span className={styles.cmMCategoryDropdown_ElementName}>{name}</MBText.Span>
        {disabled && <MBText.Span className={styles.cmMCategoryDropdown_UnavailableMessage}>&ensp;(not available)</MBText.Span>}
      </MBLink.View>
    </li>
  );
};

const CategoryElementIcon = ({category_name, icon_banner}) => {
  if (!icon_banner) return null;

  return (
    <img
      className={styles.cmMCategoryDropdown_ElementIcon}
      alt={category_name}
      src={icon_banner.image_url}
      height={icon_banner.image_height}
      width={icon_banner.image_width} />
  );
};

export default withUniqueId('content_layout_id')(withContentLayout('Web_Navigation_Mobile_Category')(CategoryDropdown));
