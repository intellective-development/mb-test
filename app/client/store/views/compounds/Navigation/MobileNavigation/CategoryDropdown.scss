@import "~client/store/views/style/variables";

$dropdown_height: 360px; // hardcoded, used to animate the dropdown
$dropdown_shadow_height: 4px;

// NOTE: that these classes are very, very similar to their counterparts in DesktopNavigation
.cmMCategoryDropdown_Wrapper {
  position: absolute;
  width: 100vw;
  left: 0;
  margin: 0;
  top: 100%; // stick it below the bottom of the container
  z-index: 10;
  border-top: $default-border;

  // this padding allows the dropdown's box shadow to show, despite the above overflow: hidden
  padding-bottom: $dropdown_shadow_height;

  overflow: hidden;
  visibility: hidden;
}
.cmMCategoryDropdown_Wrapper__DropdownVisible {
  visibility: visible;
}

.cmMCategoryDropdown {
  background: $brand-white;
  transition: all 200ms ease-in-out;
  margin-bottom: 0;
  padding: 12px 0 16px;
  box-shadow: 0 $dropdown_shadow_height/2  $dropdown_shadow_height 0 rgba(0, 0, 0, 0.33);

  transition: all 250ms ease-in; // animate out
  transform: translateY(-$dropdown_height);
  opacity: 0.5;
}
.cmMCategoryDropdown__Visible {
  transition-timing-function: ease-out; // animate in
  transform: translateY(0);
  opacity: 1;
}

.cmMCategoryDropdown_Element {
  @include font-size(24);

  display: inline-block;
  padding: 13px 24px;
  color: $brand-black;
}

.cmMCategoryDropdown_Element__Disabled {
  opacity: 0.25;
}

.cmMCategoryDropdown_ElementIcon {
  margin-right: 24px;
}

.cmMCategoryDropdown_ElementName {
  display: inline-block;
  padding-top: 4px;
  text-transform: capitalize;
}

.cmMCategoryDropdown_UnavailableMessage {
  @include font-size(13);
}