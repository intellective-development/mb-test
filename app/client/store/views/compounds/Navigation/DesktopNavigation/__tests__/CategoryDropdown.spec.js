
import React from 'react';
import content_module_factory from 'store/business/content_module/__tests__/content_module.factory';

// import TestProvider from 'store/views/__tests__/utils/TestProvider';
import CategoryDropdown from '../CategoryDropdown';

describe('CategoryDropdown', () => {
  it('renders', () => {
    const content_modules = [
      content_module_factory.build('product_type_link_list'),
      content_module_factory.build('link_list'),
      content_module_factory.build('carousel')
    ];

    expect(shallow(<CategoryDropdown content_modules={content_modules} />)).toMatchSnapshot();
  });
});
