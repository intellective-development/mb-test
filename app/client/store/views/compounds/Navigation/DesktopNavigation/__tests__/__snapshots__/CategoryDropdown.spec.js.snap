// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryDropdown renders 1`] = `
<div
  className="cmDCategoryDropdown_Wrapper"
>
  <div
    className="cmDCategoryDropdown_Dropdown"
  />
</div>
`;
